cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkAuthenticationExceptionHandler.java

```diff
@@ -57,10 +81,53 @@
         DEFAULT_ERROR_LIST.add(OneTimePasswordRequiredException.class);
     }
 
+    static {
+        // Username does not exist
+        THROTTLE_INCREASE_SET.add(javax.security.auth.login.AccountNotFoundException.class.getSimpleName());
+        // Wrong password
+        THROTTLE_INCREASE_SET.add(javax.security.auth.login.FailedLoginException.class.getSimpleName());
+        // Wrong one time password
+        THROTTLE_INCREASE_SET.add(OneTimePasswordFailedLoginException.class.getSimpleName());
+    }
+
     /**
      * The Open Science Framework Authentication Exception Handler.
      */
     public OpenScienceFrameworkAuthenticationExceptionHandler() {
         super.setErrors(DEFAULT_ERROR_LIST);
     }
+
+    /**
+     * Check if the authentication exception should trigger throttle increase.
+     *
+     * @param handleErrorName the simple name of the exception
+     * @return true if trigger, false otherwise
+     */
+    public static Boolean isTriggerThrottleIncrease(final String handleErrorName) {
+        return THROTTLE_INCREASE_SET.contains(handleErrorName);
+    }
+
+    /**
+     * The authentication exception handler event.
+     * 1. handle authentication exception
+     * 2. recognize those failures that should trigger login throttle
+     * 3. update flow scope in context
+     *
+     * @param context the request context
+     * @param e the authentication exception
+     * @param messageContext the message context
+     * @return an Event with the name of the exception
+     */
+    public Event preHandle(final RequestContext context, final AuthenticationException e, final MessageContext messageContext) {
```

Comment:
User `preHandle()` to:
- call `super.handle`
- put the name of exception in `OSFLoginContext`